src/main/java/com/billyyccc/mssqlclient/impl/codec/MSSQLDataTypeCodec.java

```diff
@@ -39,18 +50,137 @@ static String inferenceParamDefinitionByValueType(Object value) {
+      case MSSQLDataTypeId.DECIMALNTYPE_ID:
+        return decodeNumeric((NumericDataType) dataType, in);
+      case MSSQLDataTypeId.FLT4TYPE_ID:
+        return decodeFloat4(in);
+      case MSSQLDataTypeId.FLT8TYPE_ID:
+        return decodeFloat8(in);
+      case MSSQLDataTypeId.BITTYPE_ID:
+        return decodeBit(in);
+      case MSSQLDataTypeId.DATENTYPE_ID:
+        return decodeDateN(in);
+      case MSSQLDataTypeId.TIMENTYPE_ID:
+        return decodeTimeN((TimeNDataType) dataType, in);
       case MSSQLDataTypeId.BIGVARCHRTYPE_ID:
-        int length = in.readUnsignedShortLE();
-        return in.readCharSequence(length, StandardCharsets.UTF_8);
+        return decodeVarchar(in);
       default:
         throw new UnsupportedOperationException("Unsupported datatype: " + dataType);
     }
   }
+
+  private static LocalTime decodeTimeN(TimeNDataType dataType, ByteBuf in) {
+    int scale = dataType.scale();
+    byte timeLength = in.readByte();
+    long timeValue;
+    switch (timeLength) {
+      case 0:
+        return null;
+      case 3:
+        timeValue = in.readUnsignedMediumLE();
+        break;
+      case 4:
+        timeValue = in.readUnsignedIntLE();
+        break;
+      case 5:
+        timeValue = readUnsignedInt40LE(in);
+        break;
+      default:
+        throw new IllegalStateException();
+    }
+    for (int i = 0; i < 7 - scale; i++) {
+      timeValue *= 10;
+    }
+    timeValue = (long) (timeValue * Math.pow(10, 7 - scale));
+    long secondsValue = timeValue / 100000000;
+    long nanosValue = timeValue % 100000000;
+    return LocalTime.ofSecondOfDay(secondsValue).plusNanos(nanosValue);
+  }
+
+  private static CharSequence decodeVarchar(ByteBuf in) {
+    int length = in.readUnsignedShortLE();
+    return in.readCharSequence(length, StandardCharsets.UTF_8);
+  }
+
+  private static LocalDate decodeDateN(ByteBuf in) {
+    byte dateLength = in.readByte();
+    if (dateLength == 0) {
+      return null;
+    } else if (dateLength == 3) {
+      int days = in.readUnsignedMediumLE();
+      return START_DATE.plus(days, ChronoUnit.DAYS);
+    } else {
+      throw new IllegalStateException();
+    }
+  }
+
+  private static boolean decodeBit(ByteBuf in) {
+    return in.readBoolean();
+  }
+
+  private static double decodeFloat8(ByteBuf in) {
+    return in.readDoubleLE();
+  }
+
+  private static float decodeFloat4(ByteBuf in) {
+    return in.readFloatLE();
+  }
+
+  private static Numeric decodeNumeric(NumericDataType dataType, ByteBuf in) {
+    short length = in.readUnsignedByte();
+    if (length == 0) {
+      return null;
+    } else {
+      int sign = in.readByte();
+      Number value;
+      switch (length - 1) {
+        case 4:
+          value = in.readIntLE();
+          break;
+        case 8:
+          value = in.readLongLE();
+          break;
+        case 12:
+        case 16:
+          //TODO wrapped types?
+          throw new UnsupportedOperationException();
```

Comment:
Do we need a test for this ?

Reply:
updated with https://github.com/BillyYccc/incubator-vertx-mssql-client/pull/18/commits/03f2b1be16911f40974a84e4a6206048437d06e0.